### PR TITLE
Implementation of linkage algorithm against new schema

### LINKED ISSUE #23: simple_link module
## Summary
Create a new module, `linkage/simple_link.py` that replicates the functionality of the `linkage/link.py` module, but uses the new schema defined in models.py for blocking.

## Acceptance Criteria
- [ ] Implement the link_record_against_mpi function
- [ ] Implement the add_person_resource function
- [ ] Test cases for the above functions

## Background / Context
Provide any additional context, background information, or references that might be useful for understanding the issue.

## Dependencies
#10

### PR DESCRIPTION
**Pull request by ericbuckley**
## Description
This is the big one, this is a reimplementation of the linkage algorithm using the new schema.  The new algorithm has been added to `recordlinker.linking.link`.

## Related Issues
closes #23 

## Additional Notes
* Replaced the `FEATURE` literal type with a `Feature` enum for better type safety and readability. (`src/recordlinker/models/pii.py`)
* Separated matchers into two files based on the signature it was using (old schema vs new schema) (`src/recordlinker/linkage/matchers.py` and `src/recordlinker/linking/matchers.py`)
* Changed `FEATURE_COMPARE_FUNC` and `MATCH_RULE_FUNC` to use floats instead of booleans for more granular matching. (`src/recordlinker/linking/matchers.py`)
* Added caching to the Patient model for generating its record.  (`src/recordlinker/models/mpi.py`)

<--------------------- REMOVE THE LINES BELOW BEFORE MERGING --------------------->

## Checklist
Please review and complete the following checklist before submitting your pull request:

- [x] I have ensured that the pull request is of a manageable size, allowing it to be reviewed within a single session.
- [x] I have reviewed my changes to ensure they are clear, concise, and well-documented.
- [x] I have updated the documentation, if applicable.
- [x] I have added or updated test cases to cover my changes, if applicable.
- [x] I have minimized the number of reviewers to include only those essential for the review.
- [x] I have notified teammates in the review thread to build awareness.

## Checklist for Reviewers
Please review and complete the following checklist during the review process:

- [ ] The code follows best practices and conventions.
- [ ] The changes implement the desired functionality or fix the reported issue.
- [ ] The tests cover the new changes and pass successfully.
- [ ] Any potential edge cases or error scenarios have been considered.
This pull request includes significant changes to the `recordlinker` module to improve feature matching and update the codebase to use more descriptive and consistent identifiers. The most important changes involve modifying the feature comparison functions to return float values instead of booleans, updating the `Feature` model to use an enum, and adjusting related unit tests.

### CODE REVIEW

File: src/recordlinker/models/pii.py [hunk 1]
```diff
@@ -5,18 +5,27 @@
...
-    "state",
-    "zip",
-]
+
+class Feature(enum.Enum):
+    """
+    Enum for the different Patient attributes that can be used for comparison.
+    """
+
+    BIRTHDATE = "birthdate"
```
Comment (derekadombek) [inline src/recordlinker/models/pii.py:14]: should we change this to `birth_date` as well?

File: src/recordlinker/linking/link.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                # The match ratio is larger than the minimum cluster threshold,
+                # optionally update the max score for this person
+                scores[person] = max(scores[person], match_ratio)
+
+    matched_person: typing.Optional[models.Person] = None
+    if scores:
+        # Find the person with the highest matching score
+        matched_person, _ = max(scores.items(), key=lambda i: i[1])
+
+    patient = mpi_service.insert_patient(
```
Comment (derekadombek) [inline src/recordlinker/linking/link.py:173]: I think I misses this in the last PR but cool to see this in action.

File: src/recordlinker/linkage/matchers.py [hunk 3]
```diff
@@ -435,8 +435,8 @@ def _get_fuzzy_params(col: str, **kwargs) -> tuple[SIMILARITY_MEASURES, float]:
 
 
 def single_feature_match_exact(
-    record: models.PIIRecord, patient: models.Patient, key: models.FEATURE, **kwargs: dict
-) -> bool:
+    record: models.PIIRecord, patient: models.Patient, key: models.Feature, **kwargs: dict
+) -> float:
```
Comment (cbrinson-rise8) [inline src/recordlinker/linkage/matchers.py:439]: curious why this was changed to a float instead of staying as a `bool` or `int`

File: src/recordlinker/linkage/matchers.py [hunk 4]
```diff
@@ -447,18 +447,18 @@ def single_feature_match_exact(
...
+                return 1
+    return 0
 
 
-def single_feature_match_fuzzy_string(
-    record: models.PIIRecord, patient: models.Patient, key: models.FEATURE, **kwargs: dict
-) -> bool:
+def single_feature_match_fuzzy(
+    record: models.PIIRecord, patient: models.Patient, key: models.Feature, **kwargs: dict
+) -> float:
```
Comment (cbrinson-rise8) [inline src/recordlinker/linkage/matchers.py:461]: Curious why was this changed to a float instead of staying a `bool`

Comment (ericbuckley) [reply src/recordlinker/linkage/matchers.py:461]: As you can see from line 22, the signature of these functions changed to return a float rather than a bool.
```
FEATURE_COMPARE_FUNC = typing.Callable[[models.PIIRecord, models.Patient, models.Feature], float]
```

As I dug in deeper to the existing feature matching functions, I noticed that some of them (specifically the log_odds versions) were returning floats.  It just so happens, that the eval functions were also kind of expecting a list of numerics to determine if its a match or not.  They have been previously taking advantage of that fact that you can sum booleans and get a numeric value.  For example, `sum([True, False, True, False, False]) == 2` in python.

File: src/recordlinker/models/pii.py (hunk 1 above)
Comment (ericbuckley) [reply src/recordlinker/models/pii.py:14]: I think eventually that change makes sense, because we're already using snake case for `first_name` and `last_name`, however I purposefully didn't want to do that now.

My goal, is to get the new schema and code working in near parallel to the existing schema.  That means, we should be able to use the existing DIBBs basic and enhanced algorithms with both of the schemas.  Right now, that's technically not completely possible, because the matching function signatures have changed between the two schemas, but other than that I'm trying to leave the algorithm definitions intact so we need minimal changes to that config when we move back and forth between testing the old and new schema.  Because the old codebase is still referencing the feature as "birthdate" in that codebase, I choose to leave it as is for now.

File: tests/unit/test_matchers.py [hunk 5]
```diff
@@ -160,13 +161,73 @@ def test_feature_match_four_char():
 
 
 def test_feature_match_exact():
-    record_i = ["240 Rippin Ranch Apt 66", "1980-01-01", "Los Angeles", "Verónica Eva", "Reynoso", "834d436b-9e1f-2e8e-f28a-ad40bef9f365", "female", "CA", "90502"]
-    record_j = ["240 Rippin Ranch Apt 66", "1980-01-01", "Los Angeles", "Verónica Eva", "Reynoso", "834d436b-9e1f-2e8e-f28a-ad40bef9f365", "female", "CA", "90502"]
-    record_k = ["240 Rippin Ranch Apt 66", datetime.date(1980, 1, 1), "Los Angeles", "Verónica Eva", "Reynoso", "834d436b-9e1f-2e8e-f28a-ad40bef9f365", "female", "CA", "90502"]
-    record_l = ["123 X St Unit 2", "1995-06-20", "Chicago", "Alejandra", "Arenas", "124d436b-9e1f-2e8e-f28a-ad40bef9f367", "male", "IL", "60615"]
-    record_m = ["123 X St Unit 2", datetime.date(1980, 6, 20), "Chicago", "Alejandra", "Arenas", "124d436b-9e1f-2e8e-f28a-ad40bef9f367", "male", "IL", "60615"]
+    record_i = [
```
Comment (alhayward) [inline tests/unit/test_matchers.py:164]: Much more readable! Thank you 😄

File: src/recordlinker/linking/link.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
+"""
+recordlinker.linking.link
```
[the remark this answers is not in the document]
Comment (ericbuckley) [reply src/recordlinker/linking/link.py:2]: Its a temporary change so we can have two schemas at the same time (linkage == old schema code, linking == new schema code).  Long term, it will just be `recordlinker.linkage`.

File: src/recordlinker/linking/link.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+        # evaluate each Person cluster to see if the incoming record is a match
+        for person, patients in clusters.items():
+            matched_count = 0
+            for patient in patients:
+                # increment our match count if the pii_record matches the patient
+                if compare(pii_record, patient, linkage_pass):
+                    matched_count += 1
+            # calculate the match ratio for this person cluster
+            match_ratio = matched_count / len(patients)
```
[the remark this answers is not in the document]
Comment (ericbuckley) [reply src/recordlinker/linking/link.py:162]: Thanks for jogging my memory on this.  I don't want to make that change here, this PR is just about replicating the existing algorithm using the new schema and the functions built to interact with it.  I would however like to create a placeholder GH issue to do some research into that.  I just added a draft issue to our backlog to track this.  I'll add more information on Wed during the sprint planning session.  https://github.com/orgs/CDCgov/projects/34/views/1?pane=issue&itemId=81649680

File: assets/linking/basic_algorithm.json [hunk 8]
```diff
@@ -0,0 +1,68 @@
+{
+  "algorithm": [
+    {
+      "funcs": {
+        "first_name": "func:recordlinker.linkage.matchers.single_feature_match_fuzzy",
```
[the remark this answers is not in the document]
Comment (ericbuckley) [reply assets/linking/basic_algorithm.json:5]: I appreciate all the recommendations here, and I will have a change coming shortly that changes some of these again, however...

For this next phase of the project, I want to test the differences between the old and new schemas.  To make this easy, I'm going to try to make as little changes to the algorithm as possible when switching between the two.  Thus, I'm going to keep the names identical to what the DIBBs team named them originally, when applicable.  I've added a task to the board for a rename in the future once we've removed the old schema https://github.com/orgs/CDCgov/projects/34/views/1?pane=issue&itemId=81661506

File: src/recordlinker/linkage/matchers.py [hunk 9]
```diff
@@ -18,9 +18,9 @@
 SIMILARITY_MEASURES = typing.Literal["JaroWinkler", "Levenshtein", "DamerauLevenshtein"]
 
 # A Callable type for comparing a feature on two records
-FEATURE_COMPARE_FUNC = typing.Callable[[models.PIIRecord, models.Patient, models.FEATURE], bool]
+FEATURE_COMPARE_FUNC = typing.Callable[[models.PIIRecord, models.Patient, models.Feature], float]
```
[the remark this answers is not in the document]
Comment (ericbuckley) [reply src/recordlinker/linkage/matchers.py:21]: The `models.Patient` class does directly correspond to the `mpi_patient` table in the database, so I would very much like to leave those names matching, just so it's easy for people to make the connection between the model and the table.  I'm open to new names for `models.PIIRecord` though, as that's really just a serializer meant to store and capture PII data to compare.  However, the model is used by BOTH incoming and existing payloads, so naming it `models.IncomingPatient` might be confusing (like when we do something like `patient.incomingpatient.data`).  Maybe there is a better term for labeling the serialized data we want to compare?

Alternatively, we can also just rename the variables in the matcher functions themselves to be more descriptive.

**Review by cbrinson-rise8 — APPROVED** [review]

File: src/recordlinker/linkage/matchers.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (alhayward) [reply src/recordlinker/linkage/matchers.py]: Can you ELI5 the purposes of the `models.Patient` and `models.PIIRecord` classes, respectively? How `models.PIIRecord` is used for both incoming and existing payloads? I think this will help inform naming.

[the remark this answers is not in the document]
Comment (ericbuckley) [reply src/recordlinker/linkage/matchers.py]: I'll do my best....

`models.Patient` is intended to store all the Patient information we received in a previous health care document, as well as any calculations we made on the data.  It's a historical record of the document, at least the Patient information within.  A models.Patient, inherently **has** a reference to a models.PIIRecord.

`models.PIIRecord` is a lightweight structure around a collection of patient identifiers.  Having a defined structure around the information makes it easier to compare data elements between different payloads (aka PIRecords).  A models.PIIRecord object **does not know of** a models.Patient.
